---
title: 
- <center>Hardy Weinberg III <br> The Relationship between Genotype and Allele Frequencies</center>
output:
  html_document:
    css: Chapter.css
    toc: yes
    toc_depth: 5
---

### Background

Think about relationship between p and heterozygosity (fraction of individuals that are heterozygous).  At extremes - p= 0 and p=1, heterozygosity is obviously zero, but what  what happens in between? A very important point to recognize is that while the frequency of heterozygotes is a function of p, the frequency of homozygotes is a function of p^2.  What that means qualitatively is that as p gets small, p^2 gets very small, meaning that, in a finite sample, it is quite possible that there will be no homozygotes for the minor allele present.

#### Heterozygosity when q is small

We can look at this graphically as follows:  What we will do will be to plot the ration of expected heterozygotes to   aa homozygotes (where f(a)=q), or 2pq/q^2, first as p varies from 0 to .8

```{r}
p <-seq(0,.9,.001)
q <-1-p
fhet <-(2*p)/q
plot(p, fhet,type="l", ylab="heterozygote/homozygote", main="Heterozygote/Homozygote as a Function of p")
```
And we see that as p increases, the ratio increases exponentially.  This becomes even more extreme as we approach 1:
```{r}
p <-seq(.9,1,.001)
q <-1-p
fhet <-(2*p)/q
plot(p, fhet,type="l", ylab="heterozygote/homozygote", main="Heterozygote/Homozygote as a Function of p")
```
In fact, in this range, we can use the approximation of p ≈ 1, so that the ratio then ≈ 2/q:
```{r}
plot(p, fhet,type="l", ylab="heterozygote/homozygote", main="Heterozygote/Homozygote as a Function of p")
lines(p,2/q,col="red", type="l")
```
And we see that indeed the fit is quite close

#### Exploring the entire range of allele frequencies

Now we want to examine expected heterozygosity for the entire range of possible values for p (0 to 1).  We can explore this graphically, using what is called a DeFinetti diagram (or sometimes a ternary plot).  It is best to work through a few by example.  Fortunately, the R package HardyWeinberg has excellent capabilities in this regard.
```{r,message=FALSE}
library(HardyWeinberg)
```
We will start by generating a plot that shows the theoretical relationship between p and heterozygosity.  First, however, we need to generate simulated  multiple locus data; for example 100 genes in a sample of 100

```{r}
dat <-(HWData(100,100))
str(dat)
```
This returns a list containing two matrices, the first element of which, dat\$Xt, consists of genotype numbers and the second element of which, dat\$Xc, consists of genotype frequencies.  We can look at a piece of the one we are interested in, dat\$Xt, as follows:

```{r}
head(dat$Xt)
```
And its structure is pretty clear.  We will come back to using these numbers in a minute, but first, we'd like to visualize the theoretical relationship between p and heterozygosity.  We can do so as follows:


```{r,results='hide'}
gdist <-dat$Xt #create a variable with the working data
HWTernaryPlot(gdist, hwcurve=TRUE,addmarkers=FALSE,region=0,vbounds=FALSE,axis=2,vertexlab=c("0","","1"),axislab="P",main="Theoretical Relationship betweeen p and H",cex.main=1.2)
abline(h=.5,lty=2)
```
The parabola in this curve indicates the expected heterozygosity given the value of p plotted on the x axis.  What we see is that heterozygosity is maximized at 0.5 when p equals .5  (indicated by the dashed line).  

Now we can add our simulated data to the plot.  Remember that we generated 100 simulated data points (genotype distributions for 100 alleles in a sample size of 100, all drawn from expected HW distribution).  Let's put them onto the de Finetti diagram
```{r,results='hide'}
HWTernaryPlot(gdist, hwcurve=TRUE,addmarkers=TRUE,region=0,vbounds=FALSE,axis=2,vertexlab=c("0","","1"),axislab="P",main="Adding 100 Loci from N=100",cex.main=1.2)
```
And we see we have a bunch of scatter (although the theoretical curve is followed pretty well.)  But what falls within chi square acceptance value?

```{r,results='hide'}
options(warn=-1)
HWTernaryPlot(gdist, hwcurve=TRUE,addmarkers=TRUE,region=1,vbounds=FALSE,axis=2,vertexlab=c("0","","1"))
```
Remember, testing at the 5 % level, so not surprising that a few will fall outside of acceptance region.  But the key point to remember is that the data we have simulated here were 100 loci from 100 individuals *in a Hardy Weinberg Population*.  In our next exercise, we will take this one step further and look at some real data.  

<a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/deed.en_US"><img alt="Creative Commons License" style="border-width:0" src="http://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /></a><br />This work is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/deed.en_US">Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License</a>.